# Model Validation Techniques

### Train and test splits

In [2]:
import pandas as pd
from pandas import read_csv
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold 
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import LeaveOneOut
import warnings
warnings.filterwarnings("ignore")

In [4]:
diabetes = "pima-indians-diabetes.data.csv"

In [5]:
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']

In [6]:
dataframe = read_csv(diabetes,names=names)

In [7]:
array = dataframe.values

In [8]:
x = array[:,0:8]
y = array[:,8]

In [26]:
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size = 0.33,random_state = 7)

In [27]:
model = LogisticRegression()

In [28]:
model.fit(x_train,y_train)

LogisticRegression()

In [29]:
result = model.score(x_test,y_test)                         # This will accuracy of the test data and internally predict the new values from x_test data and those predicted values will be compared with the actual values(y_test)

In [30]:
result*100

78.74015748031496

Inference: Accuracy of our model is 78.74%

### KFold Cross validation technique

In [42]:
kfold = KFold(n_splits = 10, random_state = 7, shuffle = True)

In [43]:
model = LogisticRegression(max_iter = 400)

In [45]:
result = cross_val_score(model, x,y, cv = kfold)

In [46]:
result

array([0.83116883, 0.74025974, 0.74025974, 0.80519481, 0.79220779,
       0.77922078, 0.66233766, 0.80519481, 0.82894737, 0.73684211])

In [47]:
result.mean()*100

77.21633629528367

In [48]:
result.std()*100

4.96837651757489

### Leave One Out Cross Validation

In [9]:
loo = LeaveOneOut()
model = LogisticRegression(max_iter = 400)
results = cross_val_score(model, x,y, cv = loo)

In [10]:
results.mean()*100

77.60416666666666

In [12]:
results.std()*100                 # This will give the variations in accuracies among all the folds

41.68944689773287